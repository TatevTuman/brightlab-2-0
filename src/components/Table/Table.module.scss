.table {
  width: 100%;
  text-align: left;
  border-collapse: separate;
  margin-bottom: 32px;
  border-spacing: 0 4px;
}
