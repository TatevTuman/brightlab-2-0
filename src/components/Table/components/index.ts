export { default as TableBody } from './TableBody/TableBody'
export { default as TableHeader } from './TableHeader/TableHeader'
