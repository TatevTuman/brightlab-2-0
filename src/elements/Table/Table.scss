@import "static/assets/styles/mixins";

.grid-table {
  display: grid;
  border-top: 1px solid var(--purple);
  border-right: 1px solid var(--purple);

  &-column, &-cell {
    padding: toRem(8) toRem(4);
    border-bottom: 1px solid var(--purple);
    border-left: 1px solid var(--purple);
  }
}
