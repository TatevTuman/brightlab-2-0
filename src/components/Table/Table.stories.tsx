import React from 'react'
import Table from './Table'
import { Button } from '@elements'
import { TableColumnType } from '@types'

export default {
  title: 'Table',
  component: Table
}

export const Story = () => {
  const columns: TableColumnType[] = [
    {
      title: 'Имя',
      dataIndex: 'name',
      key: 'name'
    },
    {
      title: 'Фамилия',
      dataIndex: 'surname',
      key: 'surname'
    },
    {
      title: 'Возраст',
      dataIndex: 'age',
      key: 'age'
    },
    {
      title: '',
      dataIndex: 'info',
      key: 'info',
      render(cell, row) {
        const info = row.name + ' ' + row.surname + ' ' + row.age
        return (
          <Button type={'secondary'} size={'md'} onClick={() => alert(info)}>
            Get Info
          </Button>
        )
      }
    }
  ]

  const rows = new Array(10).fill(0).map(() => ({
    name: 'John',
    surname: 'Doe',
    age: '34'
  }))

  return (
    <div className="story">
      <Table columns={columns} rows={rows} handleRowClick={() => alert('Row clicked!')} isRowIndex />
    </div>
  )
}
