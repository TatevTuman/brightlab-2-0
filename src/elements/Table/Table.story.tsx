import React from 'react'
import { Meta, Story } from '@storybook/react/types-6-0'
import Table, { TableProps } from './Table'
import { storybook } from '@utils'
import { AnyObject } from '@types'

interface TableStoryTemplateProps extends TableProps<AnyObject> {}

export default {
  title: 'Elements/Table',
  argTypes: {
    columns: { control: 'object', name: 'Columns' },
    data: { control: 'object', name: 'Data' }
  }
} as Meta

const TableStoryTemplate: Story<TableStoryTemplateProps> = args => {
  return <Table {...args} />
}

const columns = [
  {
    key: 'column1',
    title: 'Column1'
  },
  {
    key: 'column2',
    title: 'Column2'
  },
  {
    key: 'column3',
    title: 'Column3'
  },
  {
    key: 'column4',
    title: 'Column4'
  },
  {
    key: 'column5',
    title: 'Column5'
  },
  {
    key: 'column6',
    title: 'Column6'
  }
]

const data = new Array(10).fill(0).map((_, index) => {
  return columns.reduce<{ [x: string]: string }>((acc, column) => {
    return {
      ...acc,
      [column.key]: 'Lorem ipsum dolores sit amet'
    }
  }, {})
})

TableStoryTemplate.args = {
  columns,
  data
}

export const TableStory = TableStoryTemplate.bind({})
TableStory.args = {
  ...TableStoryTemplate.args
}
