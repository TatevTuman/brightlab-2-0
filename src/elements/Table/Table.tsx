import React, { CSSProperties, memo } from 'react'
import { get } from 'lodash'
import './Table.scss'

export type TableColumnType<T> = {
  key: string
  title: string
  style?: CSSProperties
  render?(field: any, item: T): JSX.Element
}

export type TableCellType = {
  disabled?: boolean
}

export interface TableProps<T> {
  columns: TableColumnType<T>[]
  data: T[]
}

const TableColumns = <T extends TableCellType>(props: { columns: TableColumnType<T>[] }) => {
  const { columns } = props

  return (
    <>
      {columns.map(columns => {
        const { key, title, style } = columns

        return (
          <div key={`column-${key}`} className={'grid-table-column'} style={style}>
            {title}
          </div>
        )
      })}
    </>
  )
}

const TableRows = <T extends TableCellType>(props: TableProps<T>) => {
  const { columns, data } = props

  return (
    <>
      {data.map((item: T, index: number) => {
        return columns.map(column => {
          const { key } = column

          const itemData = get(item, key, null)
          const render = column.render ? column.render(itemData, item) : itemData

          return (
            <div key={`cell-${key}-${index}`} className={'grid-table-cell'} data-disabled={item.disabled}>
              {render}
            </div>
          )
        })
      })}
    </>
  )
}

const Table = <T extends TableCellType>(props: TableProps<T>) => {
  const { columns, data } = props

  return (
    <div className={'grid-table'} style={{ gridTemplateColumns: `repeat(${columns.length}, 1fr)` }}>
      <TableColumns<T> columns={columns} />
      <TableRows<T> columns={columns} data={data} />
    </div>
  )
}

export default memo(Table)
