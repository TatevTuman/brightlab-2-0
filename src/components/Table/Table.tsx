import React, { memo } from 'react'
import { TableColumnType, TableRowType } from '@types'
import { TableHead, TableBody } from './components'
import './Table.scss'

interface TableProps {
  columns: TableColumnType[]
  rows: TableRowType[]
  filters?: { field: string; handleFilter: (...args: any) => boolean }[]
  handleRowClick?: (row: TableRowType) => void
  isRowIndex?: boolean
  minWidth?: string
}

const Table: React.FC<TableProps> = props => {
  const { columns, rows, filters, handleRowClick, isRowIndex, minWidth } = props

  return (
    <table className={'table'} style={{ minWidth }}>
      <TableHead columns={columns} isRowIndex={isRowIndex} />
      <TableBody
        rows={rows}
        columns={columns}
        filters={filters}
        handleRowClick={handleRowClick}
        isRowIndex={isRowIndex}
      />
    </table>
  )
}

export default memo(Table)
