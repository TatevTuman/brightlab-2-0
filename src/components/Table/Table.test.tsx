import React from 'react'
import renderer from 'react-test-renderer'
import Table from './Table'
import { TableColumnType } from '@types'
import { Button } from '@elements'

describe('Table', () => {
  it('render()', () => {
    const columns: TableColumnType[] = [
      {
        title: 'Имя',
        dataIndex: 'name',
        key: 'name'
      },
      {
        title: 'Фамилия',
        dataIndex: 'surname',
        key: 'surname'
      },
      {
        title: 'Возраст',
        dataIndex: 'age',
        key: 'age'
      },
      {
        title: '',
        dataIndex: 'info',
        key: 'info',
        render(cell, row) {
          return (
            <Button type={'secondary'} size={'md'}>
              Get Info
            </Button>
          )
        }
      }
    ]

    const rows = new Array(10).fill(0).map(() => ({
      name: 'John',
      surname: 'Doe',
      age: '34'
    }))

    const table = renderer
      .create(
        <div className="story">
          <Table columns={columns} rows={rows} isRowIndex />
        </div>
      )
      .toJSON()
    expect(table).toMatchSnapshot()
  })
})
