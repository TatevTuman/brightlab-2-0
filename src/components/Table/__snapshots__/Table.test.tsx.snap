// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render() 1`] = `
<div
  className="story"
>
  <table
    className="table"
    style={
      Object {
        "minWidth": undefined,
      }
    }
  >
    <thead
      className="table-head"
    >
      <tr
        className="table-head-tr"
      >
        <td
          className="table-head-tr-td"
        />
        <th
          className="table-head-tr-th"
        >
          Имя
        </th>
        <th
          className="table-head-tr-th"
        >
          Фамилия
        </th>
        <th
          className="table-head-tr-th"
        >
          Возраст
        </th>
        <th
          className="table-head-tr-th"
        >
          
        </th>
      </tr>
    </thead>
    <tbody
      className="table-body"
    >
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            1
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            2
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            3
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            4
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            5
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            6
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            7
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            8
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            9
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
      <tr
        className="table-body-tr"
        data-disabled={false}
        onClick={[Function]}
      >
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "20px",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            10
            .
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            John
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            Doe
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            34
          </div>
        </td>
        <td
          className="table-body-tr-td"
          style={
            Object {
              "width": "unset",
            }
          }
        >
          <div
            className="table-body-tr-td__inner"
          >
            <button
              className="button button__secondary-md"
              data-disabled={false}
              onClick={[Function]}
            >
              Get Info
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
